src/webgpu/api/operation/vertex_state/index_format.spec.ts

```diff
@@ -0,0 +1,317 @@
+  .params([
+    { indexOffset: 0, _expectedShape: kSquare },
+    { indexOffset: 6, _expectedShape: kBottomLeftTriangle },
+    { indexOffset: 18, _expectedShape: kNothing },
+  ])
+  .fn(t => {
+    const { indexOffset, _expectedShape } = t.params;
+
+    // If this is written as uint16 but interpreted as uint32, it will have index 1 and 2 be both 0
+    // and render nothing.
+    // And the index buffer size - offset must be not less than the size required by triangle
+    // list, otherwise it also render nothing.
+    const indices: number[] = [1, 2, 0, 0, 0, 0, 0, 1, 3, 0];
+    const indexBuffer = t.CreateIndexBuffer(indices, 'uint16');
+    const result = t.run(indexBuffer, indices.length, 'uint16', indexOffset);
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(_expectedShape);
+    t.expectContents(result, expectedTextureValues);
+  });
+
+g.test('index_format,uint32')
+  .desc('Test rendering result of indexed draw with index format of uint32.')
+  .params([
+    { indexOffset: 0, _expectedShape: kSquare },
+    { indexOffset: 12, _expectedShape: kBottomLeftTriangle },
+    { indexOffset: 36, _expectedShape: kNothing },
+  ])
+  .fn(t => {
+    const { indexOffset, _expectedShape } = t.params;
+
+    // If this is interpreted as uint16, then it would be 0, 1, 0, ... and would draw nothing.
+    // And the index buffer size - offset must be not less than the size required by triangle
+    // list, otherwise it also render nothing.
+    const indices: number[] = [1, 2, 0, 0, 0, 0, 0, 1, 3, 0];
+    const indexBuffer = t.CreateIndexBuffer(indices, 'uint32');
+    const result = t.run(indexBuffer, indices.length, 'uint32', indexOffset);
+
+    const expectedTextureValues = t.CreateExpectedUint8Array(_expectedShape);
+    t.expectContents(result, expectedTextureValues);
+  });
+
+g.test('primitive_restart')
+  .desc('Test primitive restart with each primitive topology.')
+  .params(
+    params()
+      .combine(poptions('indexFormat', ['uint16', 'uint32'] as const))
+      .combine(kPrimitiveTopologiesForRestart)
+  )
+  .fn(t => {
+    const { indexFormat, primitiveTopology, _expectedShape } = t.params;
+
+    let indices: number[];
+    // Primitive restart should be always active with strip primitive topologies
+    // ('line-strip' or 'triangle-strip') and never active for other topologies, where
+    // the primitive restart value isn't special and should be treated as a regular index value.
+    //
+    // The value -1 gets uploaded as 0xFFFF or 0xFFFF_FFFF according to the format.
+    //
+    // An arrow '->' indicates the expected result for that primitiveTopology.
+    if (primitiveTopology === 'triangle-list') {
+      // -> triangle-list:              (0, 1, 3), (-1, 2, 1)
+      //    triangle-list with restart: (0, 1, 3), (2, 1, 0)
+      //    triangle-strip:             (0, 1, 3), (2, 1, 0), (1, 0, 0)
+      //    triangle-strip w/o restart: (0, 1, 3), (1, 3, -1), (3, -1, 2), (-1, 2, 1), (2, 1, 0), (1, 0, 0)
```

Comment:
In going through these to make sure the test were robust I realized I did need to draw the expected result for every incorrect topology. So I added it.

I found that some of the incorrect topologies are not distinguishable from the expected result, due to the position value (0,0,0,0) for index -1, which iiuc will never render anything.